lib/server-manager.ts

```diff
@@ -255,14 +255,50 @@ export class ServerManager {
     this._normalizedProjectPaths = atom.project.getDirectories().map((d) => this.normalizePath(d.getPath()))
   }
 
+  public getProjectPaths(): string[] {
+    return this._normalizedProjectPaths
+  }
+
+  /**
+   * Public: fetch the current open list of workspace folders
+   *
+   * @param getProjectPaths A method that returns the open atom projects. This is passed from {ServerManager.getProjectPaths}
+   * @returns A {Promise} containing an {Array} of {lsp.WorkspaceFolder[]} or {null} if only a single file is open in the tool.
+   */
+  public getWorkspaceFolders(): Promise<ls.WorkspaceFolder[] | null> {
+    // NOTE the method must return a Promise based on the specification
+    const projectPaths = this.getProjectPaths()
+    if (projectPaths.length === 0) {
+      // only a single file is open
+      return Promise.resolve(null)
+    } else {
+      return Promise.resolve(projectPaths.map(projectPathToWorkspaceFolder))
+    }
+  }
+
   public normalizePath(projectPath: string): string {
     return !projectPath.endsWith(path.sep) ? path.join(projectPath, path.sep) : projectPath
   }
 
   public async projectPathsChanged(projectPaths: string[]): Promise<void> {
-    const pathsSet = new Set(projectPaths.map(this.normalizePath))
-    const serversToStop = this._activeServers.filter((s) => !pathsSet.has(s.projectPath))
+    const pathsSet = projectPaths.map(this.normalizePath)
+    const pathsRemoved = this.getProjectPaths().filter((projectPath) => !pathsSet.includes(projectPath))
+
+    // send didChangeWorkspaceFolders
+    for (const activeServer of this._activeServers) {
+      activeServer.connection.didChangeWorkspaceFolders({
+        event: {
+          added: pathsSet.map(projectPathToWorkspaceFolder),
```

Comment:
Not sure if this `pathsSet` is all the current project paths or only those that are added.

Reply:
https://github.com/atom-community/atom-languageclient/pull/153/commits/4bce8629804ad6abe72493d2b66d94be62b2e4ee